Cache charset lookups to avoid synchronization

rossabaker opened the pull request:
Charset.forName caches only two entries before blocking threads.  If
abused, as property tests abuse them, this can cause a severe
bottleneck.  We can iterate over the available character sets and
their aliases, prewarm the cache, and avoid the method altogether.

bfritz commented:
For future reference, there's an open JDK bug for the only-cache-two behavior:
https://bugs.openjdk.java.net/browse/JDK-6717772

Given its age, it's not likely to be fixed soon.

rossabaker commented:
Cache is about 700 entries on my machine.

A dynamic cache would be a good solution for services that handle multiple real-world character sets.  It does not solve the problem of malicious requests iterating through fake character sets and triggering synchronized lookups.

If we made `java.nio.charset.Charset` optional on `Charset`, we would get a few benefits:
* An HTTP charset is just a token.  Nothing promises it points to a charset the JVM understands.
* Unless we're encoding or decoding, it doesn't need to be resolved.
* Users could write their own codecs based on `Charset` without regard to our resolution strategy, though they'd have to shadow ours.
* We could make a configurable `org.http4s.Charset => Option[java.nio.charset.Charset]` strategy, though I'll bet 99.999% of users aren't registering them dynamically after the app starts.
* Though maybe parameterizing the way HTTP is parsed in general is a good way around some other problems, like registering custom headers into a concurrent map.

reactormonk commented:
So a `case class Charset private (nioCharset: Either[String, NioCharset])` or pass in the header parser as a `Map[String, NioCharset]` in case of anything but the default.

rossabaker commented:
I've started experimenting with a Charset that has an optional NIO charset, which is a similar idea to the Either.  I think I like it: we can implement the cache that makes sense for all but the most exotic use cases, and those people are welcome to call `Charset.forName` at their own risk.

I'm having a hard time conceptualizing how and why a `CharsetProvider` would be registered after `org.http4s.Charset` is loaded.

rossabaker commented:
```scala
case req @ POST -> _ =>
  req.charset.fold("oops")(_.nioCharset.name)
```

I ran a quick benchmark with all `UTF-8` and by `fake-${counter}`, where `counter` is incremented on each request:

Charset | Without Patch | With Patch
-|-|-
constant UTF-8 | 70000 | 68000
fake-${counter} | 12000 | 67000

The bottleneck is real, but the implementation may hurt a bit in the common case.

Note: all benchmarks done on my MacBook, which generally sucks at benchmarks.

rossabaker commented:
So, I [benchmarked Charset directly](https://gist.github.com/rossabaker/eee7590fabaec1c9b6f14c3542244bbe).

```
[info] NioCharsetBench.cachedInvalid   thrpt   40    2675162.855 ±   103986.224  ops/s
[info] NioCharsetBench.cachedValid     thrpt   40   83610875.726 ±  2972073.798  ops/s
[info] NioCharsetBench.forNameInvalid  thrpt   40      34237.259 ±      783.657  ops/s
[info] NioCharsetBench.forNameValid    thrpt   40  433837293.317 ± 13169507.210  ops/s
```

This makes valid lookups 5 times worse to make invalid ones 78 times better.

rossabaker commented:
I added a benchmark, held my nose, and switched to a `java.util.HashMap`.

Costs:
- Cache init at startup.  I haven't noticed.
- 28 nanoseconds to parse charsets (on my machine).
- Can't register `CharsetProvider` after `org.http4s.Charset` is loaded.  I don't know how to do this, but I'll bet it was done once, in 2003, on WebSphere, for a since extinct dialect of Klingon.

Benefits:
- Can't lock the request pool by repeatedly POSTing `Content-Type: text/plain; charset=i-am-a-jerkface`
- Less allocations of `Charset`

We'd be [in company with Tomcat](https://github.com/apache/tomcat/blob/f89606868a44c538d15fe2d92ed45db89d5cf41f/java/org/apache/tomcat/util/buf/B2CConverter.java#L48-L55) with this approach.
